libsplinter/src/service/network/interconnect/mod.rs

```diff
@@ -13,3 +13,707 @@
 // limitations under the License.
 
 mod error;
+mod scml;
+
+use std::collections::HashMap;
+use std::sync::mpsc::{channel, Receiver, Sender};
+use std::thread;
+
+use crate::matrix::{MatrixReceiver, MatrixRecvError, MatrixSender};
+use crate::network::dispatch::{
+    ConnectionId, DispatchLoopBuilder, DispatchLoopShutdownSignaler, DispatchMessageSender,
+    Dispatcher, MessageSender,
+};
+use crate::protos::component::{ComponentMessage, ComponentMessageType};
+
+pub use self::error::{ServiceInterconnectError, ServiceLookupError};
+
+/// The ServiceLookup trait provides an interface for looking up details about individual service
+/// connections.
+pub trait ServiceLookup: Send {
+    /// Retrieves the connection id for a given service id, if found.
+    ///
+    /// # Errors
+    ///
+    /// Returns a ServiceLookupError if the connection id cannot be retrieved.
+    fn connection_id(&self, service_id: &str) -> Result<Option<String>, ServiceLookupError>;
+
+    /// Retrieves the service id for a given connection id, if found.
+    ///
+    /// # Errors
+    ///
+    /// Returns a ServiceLookupError if the service id cannot be retrieved.
+    fn service_id(&self, connection_id: &str) -> Result<Option<String>, ServiceLookupError>;
+}
+
+pub trait ServiceLookupProvider {
+    fn service_lookup(&self) -> Box<dyn ServiceLookup>;
+}
+
+/// ServiceInterconnect will receive incoming messages from services and dispatch them to the
+/// ComponentMessageType handlers. It will also receive messages from handlers that need to be
+/// sent to components.
+///
+/// When an incoming message is received, the connection_id is converted to a service processor
+/// identity. The reverse is done for an outgoing message.
+pub struct ServiceInterconnect {
+    dispatched_sender: Sender<SendRequest>,
+    recv_join_handle: thread::JoinHandle<()>,
+    send_join_handle: thread::JoinHandle<()>,
+    shutdown_handle: ShutdownHandle,
+}
+
+impl ServiceInterconnect {
+    /// get a new MessageSender over connections that can be used to send messages to services.
```

Comment:
The doc comments for these methods are inconsistent with capitalization and imperative/declarative mood.

Reply:
Updating for consistancy